src/VisualStudio/Core/Impl/ProjectSystem/CPS/CPSProjectFactory.cs

```diff
@@ -22,56 +21,67 @@ namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.C
             VisualStudioWorkspaceImpl workspace,
-            IProjectCodeModelFactory projectCodeModelFactory)
+            IProjectCodeModelFactory projectCodeModelFactory,
+            SVsServiceProvider serviceProvider)
         {
+            _threadingContext = threadingContext;
             _projectFactory = projectFactory;
             _workspace = workspace;
             _projectCodeModelFactory = projectCodeModelFactory;
+            _serviceProvider = (Shell.IAsyncServiceProvider)serviceProvider;
         }
 
-        IWorkspaceProjectContext IWorkspaceProjectContextFactory.CreateProjectContext(
+        IWorkspaceProjectContext IWorkspaceProjectContextFactory.CreateProjectContext(string languageName, string projectUniqueName, string projectFilePath, Guid projectGuid, object hierarchy, string binOutputPath)
+        {
+            return _threadingContext.JoinableTaskFactory.Run(() =>
+                this.CreateProjectContextAsync(languageName, projectUniqueName, projectFilePath, projectGuid, hierarchy, binOutputPath, CancellationToken.None));
+        }
+
+        public async Task<IWorkspaceProjectContext> CreateProjectContextAsync(
             string languageName,
             string projectUniqueName,
             string projectFilePath,
             Guid projectGuid,
             object hierarchy,
-            string binOutputPath)
-        {
-            var visualStudioProject = CreateVisualStudioProject(languageName, projectUniqueName, projectFilePath, hierarchy as IVsHierarchy, projectGuid);
-            return new CPSProject(visualStudioProject, _workspace, _projectCodeModelFactory, projectGuid, binOutputPath);
-        }
-
-        private VisualStudioProject CreateVisualStudioProject(string languageName, string projectUniqueName, string projectFilePath, IVsHierarchy hierarchy, Guid projectGuid)
+            string binOutputPath,
+            CancellationToken cancellationToken)
         {
             var creationInfo = new VisualStudioProjectCreationInfo
             {
                 FilePath = projectFilePath,
-                Hierarchy = hierarchy,
+                Hierarchy = hierarchy as IVsHierarchy,
                 ProjectGuid = projectGuid,
             };
 
-            var visualStudioProject = _projectFactory.CreateAndAddToWorkspace(projectUniqueName, languageName, creationInfo);
+            var visualStudioProject = await _projectFactory.CreateAndAddToWorkspaceAsync(projectUniqueName, languageName, creationInfo, cancellationToken).ConfigureAwait(false);
+
+            // At this point we've mutated the workspace.  So we're no longer cancellable.
+            cancellationToken = default;
 
             if (languageName == LanguageNames.FSharp)
             {
-                var shell = (IVsShell)ServiceProvider.GlobalProvider.GetService(typeof(SVsShell));
+                var shell = await _serviceProvider.GetServiceAsync<SVsShell, IVsShell7>().ConfigureAwait(false);
```

Comment:
❗ `IVsShell7` is a COM interface, so it needs to be invoked from the main thread.

```suggestion
                var shell = await _serviceProvider.GetServiceAsync<SVsShell, IVsShell7>().ConfigureAwait(true);
```

Reply:
doesn't that defeat teh purpose of LoadPackageAsync?  @jasonmalinowski ?

Reply:
I think the idea is it still allows a package that can do it's initialization off the UI thread to then not block the UI thread. Consider cases like if a package wants to go load other resources with I/O or compute something else.

@ryanmolden could best speak to intent here.

Reply:
ok.  but do i need to be on hte UI thread to call LoadPackageAsync?  Or can i call that from a BG thread?  i would think if it needed to be on the UI thread, it would switch to that itself

Reply:
seems totally crazy to me.  but ok.